File: MCP服务器/security/ThirdKeyAI_SchemaPin_e5ce0c8_python/python/schemapin/resolver.py
"""Discovery resolver abstraction for SchemaPin."""

import json
import os
from abc import ABC, abstractmethod
from typing import Any, Dict, List, Optional

from .bundle import SchemaPinTrustBundle
from .discovery import PublicKeyDiscovery
from .revocation import RevocationDocument, fetch_revocation_document


class SchemaResolver(ABC):
    """Abstract base class for discovery resolution."""

    @abstractmethod
    def resolve_discovery(self, domain: str) -> Optional[Dict[str, Any]]:
        """Resolve a well-known discovery document for a domain.

        Returns:
            Well-known response dict, or None if not found.
        """

    @abstractmethod
    def resolve_revocation(
        self, domain: str, discovery: Dict[str, Any]
    ) -> Optional[RevocationDocument]:
        """Resolve a revocation document for a domain.

        Returns:
            RevocationDocument or None.
        """


class WellKnownResolver(SchemaResolver):
    """Resolves discovery via standard .well-known HTTPS endpoints."""

    def __init__(self, timeout: int = 10):
        self._timeout = timeout

    def resolve_discovery(self, domain: str) -> Optional[Dict[str, Any]]:
        """Fetch discovery from .well-known endpoint."""
        try:
            return PublicKeyDiscovery.fetch_well_known(
                domain, timeout=self._timeout
            )
        except Exception:
            return None

    def resolve_revocation(
        self, domain: str, discovery: Dict[str, Any]
    ) -> Optional[RevocationDocument]:
        """Fetch revocation from the discovery's revocation_endpoint."""
        endpoint = discovery.get("revocation_endpoint")
        if not endpoint:
            return None
        return fetch_revocation_document(endpoint, timeout=self._timeout)


class LocalFileResolver(SchemaResolver):
    """Resolves discovery from local JSON files."""

    def __init__(
        self, discovery_dir: str, revocation_dir: Optional[str] = None
    ):
        self._discovery_dir = discovery_dir
        self._revocation_dir = revocation_dir

    def resolve_discovery(self, domain: str) -> Optional[Dict[str, Any]]:
        """Read {domain}.json from the discovery directory."""
        path = os.path.join(self._discovery_dir, f"{domain}.json")
        try:
            with open(path) as f:
                return json.load(f)
        except (OSError, json.JSONDecodeError):
            return None

    def resolve_revocation(
        self, domain: str, discovery: Dict[str, Any]
    ) -> Optional[RevocationDocument]:
        """Read {domain}.revocations.json from the revocation directory."""
        if not self._revocation_dir:
            return None
        path = os.path.join(
            self._revocation_dir, f"{domain}.revocations.json"
        )
        try:
            with open(path) as f:
                data = json.load(f)
                return RevocationDocument.from_dict(data)
        except (OSError, json.JSONDecodeError):
            return None


class TrustBundleResolver(SchemaResolver):
    """Resolves discovery from an in-memory trust bundle."""

    def __init__(self, bundle: SchemaPinTrustBundle):
        self._bundle = bundle

    @classmethod
    def from_json(cls, json_str: str) -> "TrustBundleResolver":
        """Create resolver from a JSON trust bundle string."""
        bundle = SchemaPinTrustBundle.from_json(json_str)
        return cls(bundle)

    def resolve_discovery(self, domain: str) -> Optional[Dict[str, Any]]:
        """Look up discovery in the bundle."""
        return self._bundle.find_discovery(domain)

    def resolve_revocation(
        self, domain: str, discovery: Dict[str, Any]
    ) -> Optional[RevocationDocument]:
        """Look up revocation in the bundle."""
        return self._bundle.find_revocation(domain)


class ChainResolver(SchemaResolver):
    """Tries multiple resolvers in order, returning the first success."""

    def __init__(self, resolvers: List[SchemaResolver]):
        self._resolvers = resolvers

    def resolve_discovery(self, domain: str) -> Optional[Dict[str, Any]]:
        """Try each resolver in order."""
        for resolver in self._resolvers:
            result = resolver.resolve_discovery(domain)
            if result is not None:
                return result
        return None

    def resolve_revocation(
        self, domain: str, discovery: Dict[str, Any]
    ) -> Optional[RevocationDocument]:
        """Try each resolver in order."""
        for resolver in self._resolvers:
            result = resolver.resolve_revocation(domain, discovery)
            if result is not None:
                return result
        return None

File: MCP服务器/security/ThirdKeyAI_SchemaPin_e5ce0c8_python/javascript/src/resolver.js
/**
 * Discovery resolver abstraction for SchemaPin.
 */

import { readFileSync } from 'node:fs';
import { join } from 'node:path';

import { PublicKeyDiscovery } from './discovery.js';
import { fetchRevocationDocument } from './revocation.js';
import { findDiscovery, findRevocation, parseTrustBundle } from './bundle.js';

/**
 * Abstract base class for discovery resolution.
 */
export class SchemaResolver {
    /**
     * Resolve a well-known discovery document for a domain.
     *
     * @param {string} domain
     * @returns {Promise<Object|null>}
     */
    async resolveDiscovery(domain) {
        throw new Error('resolveDiscovery not implemented');
    }

    /**
     * Resolve a revocation document for a domain.
     *
     * @param {string} domain
     * @param {Object} discovery
     * @returns {Promise<Object|null>}
     */
    async resolveRevocation(domain, discovery) {
        throw new Error('resolveRevocation not implemented');
    }
}

/**
 * Resolves discovery via standard .well-known HTTPS endpoints.
 */
export class WellKnownResolver extends SchemaResolver {
    constructor(timeout = 10000) {
        super();
        this._timeout = timeout;
    }

    async resolveDiscovery(domain) {
        try {
            return await PublicKeyDiscovery.fetchWellKnown(domain, this._timeout);
        } catch {
            return null;
        }
    }

    async resolveRevocation(domain, discovery) {
        const endpoint = discovery?.revocation_endpoint;
        if (!endpoint) return null;
        return await fetchRevocationDocument(endpoint, this._timeout);
    }
}

/**
 * Resolves discovery from local JSON files.
 */
export class LocalFileResolver extends SchemaResolver {
    constructor(discoveryDir, revocationDir = null) {
        super();
        this._discoveryDir = discoveryDir;
        this._revocationDir = revocationDir;
    }

    async resolveDiscovery(domain) {
        try {
            const path = join(this._discoveryDir, `${domain}.json`);
            const data = readFileSync(path, 'utf-8');
            return JSON.parse(data);
        } catch {
            return null;
        }
    }

    async resolveRevocation(domain, discovery) {
        if (!this._revocationDir) return null;
        try {
            const path = join(this._revocationDir, `${domain}.revocations.json`);
            const data = readFileSync(path, 'utf-8');
            return JSON.parse(data);
        } catch {
            return null;
        }
    }
}

/**
 * Resolves discovery from an in-memory trust bundle.
 */
export class TrustBundleResolver extends SchemaResolver {
    constructor(bundle) {
        super();
        this._bundle = bundle;
    }

    static fromJson(jsonStr) {
        const bundle = parseTrustBundle(jsonStr);
        return new TrustBundleResolver(bundle);
    }

    async resolveDiscovery(domain) {
        return findDiscovery(this._bundle, domain);
    }

    async resolveRevocation(domain, discovery) {
        return findRevocation(this._bundle, domain);
    }
}

/**
 * Tries multiple resolvers in order, returning the first success.
 */
export class ChainResolver extends SchemaResolver {
    constructor(resolvers) {
        super();
        this._resolvers = resolvers;
    }

    async resolveDiscovery(domain) {
        for (const resolver of this._resolvers) {
            const result = await resolver.resolveDiscovery(domain);
            if (result !== null) return result;
        }
        return null;
    }

    async resolveRevocation(domain, discovery) {
        for (const resolver of this._resolvers) {
            const result = await resolver.resolveRevocation(domain, discovery);
            if (result !== null) return result;
        }
        return null;
    }
}
